arch/xtensa/core/mpu.c

```diff
@@ -0,0 +1,703 @@
+		while (read_idx >= first_enabled_idx) {
+			struct xtensa_mpu_entry *entry_rd = &entries[read_idx];
+
+			if (entry_rd->as.p.mbz != 1U) {
+				struct xtensa_mpu_entry *entry_wr;
+
+				write_idx--;
+				entry_wr = &entries[write_idx];
+
+				*entry_wr = *entry_rd;
+				entry_wr->at.p.segment = write_idx;
+			}
+
+			read_idx--;
+		}
+
+		/* New first enabled entry is where the last written entry is. */
+		new_first = write_idx;
+
+		for (idx_0 = 0; idx_0 < new_first; idx_0++) {
+			struct xtensa_mpu_entry *e = &entries[idx_0];
+
+			/* Shortcut to zero out address and enabled bit. */
+			e->as.raw = 0U;
+
+			/* Segment value must correspond to the index. */
+			e->at.p.segment = idx_0;
+
+			/* No access at all for both kernel and user modes. */
+			e->at.p.access_rights =	XTENSA_MPU_ACCESS_P_NA_U_NA;
+
+			/* Use default memory type for disabled entries. */
+			e->at.p.memory_type = CONFIG_XTENSA_MPU_DEFAULT_MEM_TYPE;
+		}
+	} else {
+		/* No need to conlidate entries. Map is same as before. */
+		new_first = first_enabled_idx;
+	}
+
+	return new_first;
+}
+
+/**
+ * Add a memory region to the MPU map.
+ *
+ * This adds a memory region to the MPU map, by setting the appropriate
+ * start and end entries. This may re-use existing entries or add new
+ * entries to the map.
+ *
+ * @param[in,out] map Pointer to MPU map.
+ * @param[in] start_addr Start address of the region.
+ * @param[in] end_addr End address of the region.
+ * @param[in] access_rights Access rights of this region.
+ * @param[in] memory_type Memory type of this region.
+ * @param[out] first_idx Return index of first enabled entry if not NULL.
+ *
+ * @retval 0 Successful in adding the region.
+ * @retval -EINVAL Invalid values in function arguments.
+ */
+static int mpu_map_region_add(struct xtensa_mpu_map *map,
+			      uintptr_t start_addr, uintptr_t end_addr,
+			      uint32_t access_rights, uint32_t memory_type,
+			      uint8_t *first_idx)
+{
+	int ret;
+	bool exact_s, exact_e;
+	uint8_t idx_s, idx_e, first_enabled_idx;
+	struct xtensa_mpu_entry *entry_slot_s, *entry_slot_e, prev_entry;
+
+	struct xtensa_mpu_entry *entries = map->entries;
+
+	if (start_addr >= end_addr) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	first_enabled_idx = find_first_enabled_entry(entries);
+	if (first_enabled_idx >= XTENSA_MPU_NUM_ENTRIES) {
+
+		/*
+		 * If the last entry in the map is not enabled and the start
+		 * address is NULL, we can assume the map has not been populated
+		 * at all. This is because we group all enabled entries at
+		 * the end of map.
+		 */
+		struct xtensa_mpu_entry *last_entry = &entries[XTENSA_MPU_NUM_ENTRIES - 1];
+
+		if (!xtensa_mpu_entry_enable_get(last_entry) &&
+		    (xtensa_mpu_entry_start_address_get(last_entry) == 0U)) {
+			/* Empty table, so populate the entries as-is. */
+			if (end_addr == 0xFFFFFFFFU) {
+				/*
+				 * Region goes to end of memory, so only need to
+				 * program one entry.
```

Comment:
Seems like consolidate_entries() would handle this case, no?  Can probably remove this case from the if/else, I think?

Reply:
`consolidate_entries()` does not remove entry with address "0xFFFFFFFFU". Besides, once the address is written into the struct field, the lower 5 bits are masked off, so it would appear as valid address.